xarray/core/dataset.py

```diff
@@ -4986,7 +4986,8 @@ def reduce(
                     variables[name] = var
             else:
                 if (
-                    not numeric_only
+                    not reduce_dims
```

Comment:
actually, wouldn't it make more sense to add unreduced variables without calling `reduce`? Something like
```python
if not reduce_dims:
    variables[name] = var
elif not numeric_only or np.issubdtype(var.dtype, nb.number) or var.dtype == np.bool_:
    # ...
    variables[name] = var.reduce(...)
```

Reply:
That was my first idea, but a few test failed, so it would be a breaking change. I think it mainly has do to with how attributes are handled by reduce.

Reply:
well, `attrs` seems to be one of the issues (but one we could fix!). However, some reduction functions (e.g. `std`, `var`) need to run on variables that don't have the reduce dims, so my suggestion is actually wrong. See also the concept of `invariant_0d` introduced in #5207.

As I'm sure I will have forgotten about this in a few months: could you add a comment explaining this?

Reply:
Done!

Reply:
Great, thanks!

Reply:
last question: if I'm understanding this correctly, `not numeric_only` and all following lines will be overridden by `not reduce_dims`. Should we move that three lines down, below the `dtype` checks?